db/compaction/compaction_outputs.cc

```diff
@@ -190,13 +190,60 @@ bool CompactionOutputs::ShouldStopBefore(const CompactionIterator& c_iter) {
   // etc.
   const Slice& internal_key = c_iter.key();
   const uint64_t previous_overlapped_bytes = grandparent_overlapped_bytes_;
-  size_t num_grandparent_boundaries_crossed =
-      UpdateGrandparentBoundaryInfo(internal_key);
+  const InternalKeyComparator* icmp =
+      &compaction_->column_family_data()->internal_comparator();
+  // if compaction_->output_level() == 0, there is no need to update grandparent
+  // info, and that `grandparent` should be empty.
+  size_t num_grandparent_boundaries_crossed = 0;
+  bool should_stop_for_ttl = false;
+  if (compaction_->output_level() > 0) {
+    num_grandparent_boundaries_crossed =
+        UpdateGrandparentBoundaryInfo(internal_key);
+    // TTL state should be updated for each key
+    if (!files_to_cut_for_ttl_.empty()) {
+      if (cur_files_to_cut_for_ttl_ != -1) {
+        // Previous key is inside the range of a file
+        if (icmp->Compare(internal_key,
+                          files_to_cut_for_ttl_[cur_files_to_cut_for_ttl_]
+                              ->largest.Encode()) > 0) {
+          next_files_to_cut_for_ttl_ = cur_files_to_cut_for_ttl_ + 1;
+          cur_files_to_cut_for_ttl_ = -1;
+          should_stop_for_ttl = true;
+        }
+      } else {
+        // Look for the key position
+        while (next_files_to_cut_for_ttl_ <
+               static_cast<int>(files_to_cut_for_ttl_.size())) {
+          if (icmp->Compare(internal_key,
+                            files_to_cut_for_ttl_[next_files_to_cut_for_ttl_]
+                                ->smallest.Encode()) >= 0) {
+            if (icmp->Compare(internal_key,
+                              files_to_cut_for_ttl_[next_files_to_cut_for_ttl_]
+                                  ->largest.Encode()) <= 0) {
+              // With in the current file
+              cur_files_to_cut_for_ttl_ = next_files_to_cut_for_ttl_;
+              should_stop_for_ttl = true;
+              break;
+            }
+            // Beyond the current file
+            next_files_to_cut_for_ttl_++;
+          } else {
+            // Still fall into the gap
+            break;
+          }
+        }
+      }
+    }
+  }
 
   if (!HasBuilder()) {
     return false;
   }
 
+  if (should_stop_for_ttl) {
+    return true;
+  }
```

Comment:
Is this "if" being moved earlier mainly for saving computation when `should_stop_for_ttl==true` (I am not objecting this by the way). It's just that I moved it to right before the bottom `return false` (like previous behavior) and your test still passes 
```
    if (compaction_->immutable_options()->compaction_style ==
            kCompactionStyleLevel &&
        compaction_->immutable_options()->level_compaction_dynamic_file_size &&
        current_output_file_size_ >=
            ((compaction_->target_output_file_size() + 99) / 100) *
                (50 + std::min(grandparent_boundary_switched_num_ * 5,
                               size_t{40}))) {
      return true;
    }
  }

  if (should_stop_for_ttl) {
    return true;
  }

  return false;
```

Correct me if i'm wrong but is it the "update TTL states for all keys" essentially solved the "unnecessary file cutting"? From this aspect (not computation saving perpsecitve), moving  `if (should_stop_for_ttl) {` earlier is not needed. I am asking just because the title "Consider TTL compaction file cutting earlier ..." made me initially think moving "if (should_stop_for_ttl) {" earlier is critically needed.

Reply:
The behavior before this PR is to return directly without using a `should_stop_for_ttl` flag, since TTL was the last thing considered. Now I cannot do that since even if TTL suggests to return true, if !HasBuilder(), it should still return false. I could not think of a cleaner way to do it. Saving computation is a nice side effect. 

> is it the "update TTL states for all keys" essentially solved the "unnecessary file cutting"?

Correct, moving `if (should_stop_for_ttl)` is not needed for correctness (since all the early returns in this function returns true, except for the one checking !HasBuilder()).

Reply:
Thanks! I could see how should_stop_for_ttl is needed to work around with `!HasBuilder()` - was just confirming the benefit it brought to put `should_stop_for_ttl` earlier instead of putting it at the bottom and got my answer now thanks!